### ISSUE TITLE
Section 23.4.3 Code returns error

### ISSUE THREAD

**Issue by bfletcherjohnson** [hadley/r4ds#1374]
Code in Section 23.4.3, 2nd code box returns error, nearly immediately.  I think the problem is that cells in the ISBN column are empty until row 83240, that cell contains the ISBN for the book in that row.  open_dataset() assigns ISBN as null based on the first few thousand rows; when subsequently write_dataset encounters an number in ISBN, it throws an error.

Code in 2nd code box

```r
seattle_csv |>
  group_by(CheckoutYear) |>
  write_dataset(path = pq_path, format = "parquet")
```

returns error as shown below:

```r
> seattle_csv |>
+   group_by(CheckoutYear) |>
+   write_dataset(path = pq_path, format = "parquet")
Error: Invalid: In CSV column #7: Row #83240: CSV conversion error to null: invalid value '9781504752848'
```

There seems to be an error in csv file, column #7, row #83240. I hesitate to open a 9GB csv file

I deleted the directory and file and downloaded the file a second time; I get the same error.
Code in the previous section 23.3 works as described, suggesting that seattle.csv is NOT corrupt.

When I open the csv file in excel (excel, of course, complains that it is too big), in row 83240, column H (which is column #8 not #7) there is the number 9781504752848.  Column H is "ISBN" and appears to be empty (NA) for nearly all rows.  However, I  see that row 150026, column H contains 9781844678433, row 297216 contains 9781459234451.  These numbers are the correct ISBN for the book in the row; it is just that for the vast majority of books an ISBN is not given.

You mention that open_dataset() scans a few thousand rows to figure out the structure of the dataset.  Unfortunately, for the first few thousand rows the ISBN column is indeed null.

I am running :
RStudio 2022.12.0+353 "Elsbeth Geranium" Release (7d165dcfc1b6d300eb247738db2c7076234f6ef0, 2022-12-03) for Windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) RStudio/2022.12.0+353 Chrome/102.0.5005.167 Electron/19.1.3 Safari/537.36

**Comment by hadley** [comment]
You'll need the latest version of arrow.

**Comment by bfletcherjohnson** [comment]
Hmm, I think I am running the latest version of arrow, arrow_11.0.0.3.  I also upgraded R to R version 4.2.3 (2023-03-15 ucrt).  Still getting that error.  When I enter seattle_csv |> glimpse(), the column ISBN is type <null>.  In fact, in row 83240, there is an ISBN number entered '9781504752848'.
When I try to run the code in the 2nd code box; I get the error.

**Comment by bfletcherjohnson** [comment]
I'm sorry, the column ISBN type is null.

**Comment by hadley** [comment]
I saw that error with a previous version of arrow, but it went away so I removed the code I had used to work around it. Let me investigate a bit more.

**Comment by rkb965** [comment]
For anyone else finding there way here, this is Hadley's previous fix which seems to still work:

**Comment by rkb965** [comment]
For anyone else finding their way here, this is Hadley's previous fix which seems to work for me:

https://github.com/hadley/r4ds/blob/e5dc792c82776e8f309ccd79311e58c4d7793834/arrow.qmd#L78-L88

**Comment by thisisnic** [comment]
Is there anything I can do to help here? 

I'm planning on writing some content for the posit::conf workshop on "how to deal with these issues that can come up with large datasets" and explaining why it's happening and workarounds.  In terms of a simple and short-term solution, I can look into providing an updated version of the dataset which doesn't hit this bit of friction? I can see that "change the dataset" isn't necessarily the best solution though.

Alternatively, I've done some work to improve the API so users don't have to type horrible stuff like ` opts <- CsvConvertOptions$create(col_types = schema(ISBN = string())) ` 😱    I've created now wrappers around `open_dataset()` for CSV writing (i.e. `open_csv_dataset()` is designed to copy, where possible, the parameters for `read_csv_arrow()`), so the original read-in code could be updated to this:

```
seattle_csv <- open_csv_dataset(
  sources = "data/seattle-library-checkouts.csv", 
  col_types = schema(ISBN = string())
)
```

Happy to submit a PR - let me know.

**Comment by hadley** [comment]
That new code would be great!

**Comment by mine-cetinkaya-rundel** [comment]
This should be addressed with #1533 that just got merged.
